pkg/controller/hibernation/azure_actuator.go

```diff
@@ -0,0 +1,231 @@
+	azureClient, err := a.azureClientFn(cd, c, logger)
+	if err != nil {
+		return false, err
+	}
+	machines, err := listAzureMachines(cd, azureClient, azureNotRunningStates, logger)
+	if err != nil {
+		return false, err
+	}
+	return len(machines) == 0, nil
+}
+
+// MachinesStopped will return true if the machines associated with the given
+// ClusterDeployment are in a stopped state.
+func (a *azureActuator) MachinesStopped(cd *hivev1.ClusterDeployment, c client.Client, logger log.FieldLogger) (bool, error) {
+	logger = logger.WithField("cloud", "azure")
+	azureClient, err := a.azureClientFn(cd, c, logger)
+	if err != nil {
+		return false, err
+	}
+	machines, err := listAzureMachines(cd, azureClient, azureNotStoppedStates, logger)
+	if err != nil {
+		return false, err
+	}
+	return len(machines) == 0, nil
+}
+
+type azureMachineLister struct {
+	client        azureclient.Client
+	resourceGroup string
+	states        sets.String
+	done          chan struct{}
+	err           chan error
+	logger        log.FieldLogger
+}
+
+func listAzureMachines(cd *hivev1.ClusterDeployment, azureClient azureclient.Client, states sets.String, logger log.FieldLogger) ([]compute.VirtualMachine, error) {
+	page, err := azureClient.ListAllVirtualMachines(context.TODO(), "true")
+	if err != nil {
+		return nil, err
+	}
+	result := []compute.VirtualMachine{}
+	for page.NotDone() {
+		result = append(result, filterByResourceGroupAndState(page.Values(), clusterDeploymentResourceGroup(cd), states, logger)...)
+		if err = page.Next(); err != nil {
+			return nil, err
+		}
+	}
+	return result, nil
+}
+
+func filterByResourceGroupAndState(machines []compute.VirtualMachine, resourceGroup string, states sets.String, logger log.FieldLogger) []compute.VirtualMachine {
+	result := []compute.VirtualMachine{}
+	for _, vm := range machines {
+		if vm.ID == nil {
+			continue
+		}
+		resource, err := azure.ParseResourceID(*vm.ID)
+		if err != nil {
+			logger.WithError(err).Warningf("Failed to parse resource ID")
+			continue
+		}
+		logger.Infof("Parsed resource: %#v", resource)
+		if strings.ToUpper(resource.ResourceGroup) != strings.ToUpper(resourceGroup) {
+			continue
+		}
+		state := azureMachinePowerState(vm)
+		logger.Infof("Machine power state: %s", state)
+		if !states.Has(state) {
+			continue
+		}
+		result = append(result, vm)
+	}
+	return result
+}
+
+func azureMachinePowerState(vm compute.VirtualMachine) string {
+	if vm.InstanceView == nil || vm.InstanceView.Statuses == nil {
+		return azureUnknownPowerState
+	}
+	for _, s := range *vm.InstanceView.Statuses {
+		if s.Code != nil && strings.HasPrefix(*s.Code, azurePowerStatePrefix) {
+			return strings.TrimPrefix(*s.Code, azurePowerStatePrefix)
+		}
+	}
+	return azureUnknownPowerState
+}
+
+func clusterDeploymentResourceGroup(cd *hivev1.ClusterDeployment) string {
+	if cd.Spec.ClusterMetadata == nil {
+		return ""
+	}
+	return fmt.Sprintf("%s-rg", cd.Spec.ClusterMetadata.InfraID)
```

Comment:
What is your guess on which OpenShift version changes the name of the resource group and breaks us?

Reply:
This is already broken. The OpenShift installer allows the user to specify an existing resource group to use now.
https://github.com/openshift/installer/blob/dbf10881b664be15ba525b9be3562dc478910631/pkg/types/azure/platform.go#L69-L77

Reply:
Custom resource groups can be dealt with as part of a fix for https://issues.redhat.com/browse/CO-1150.